sharding/indexHashedNodesCoordinator.go

```diff
@@ -503,31 +519,73 @@ func (ihgs *indexHashedNodesCoordinator) EpochStartPrepare(metaHeader data.Heade
 		log.Error("saving nodes coordinator config failed", "error", err.Error())
 	}
 
-	displayNodesConfiguration(eligibleMap, waitingMap, leaving, stillRemaining, nodesConfig.nbShards)
+	displayNodesConfiguration(eligibleMap, waitingMap, newNodesConfig.leavingList, stillRemaining, newNodesConfig.nbShards)
 
 	ihgs.mutSavedStateKey.Lock()
 	ihgs.savedStateKey = randomness
 	ihgs.mutSavedStateKey.Unlock()
 }
 
-// ComputeActuallyLeaving returns the list of those nodes which are actually leaving
-func ComputeActuallyLeaving(leaving []Validator, stillRemaining []Validator) []Validator {
-	actualLeaving := make([]Validator, 0)
-	for _, shouldLeave := range leaving {
-		willRemain := false
-		for _, remains := range stillRemaining {
-			if bytes.Equal(shouldLeave.PubKey(), remains.PubKey()) {
-				willRemain = true
-				break
-			}
+// GetChance will return default chance
+func (ihgs *indexHashedNodesCoordinator) GetChance(_ uint32) uint32 {
+	return defaultSelectionChances
+}
+
+func (ihgs *indexHashedNodesCoordinator) computeNodesConfigFromList(
+	validatorInfos []*state.ShardValidatorInfo,
+) (*epochNodesConfig, error) {
+
+	leaving, err := ihgs.nodesPerShardSetter.ComputeLeaving(validatorInfos)
+	if err != nil {
+		return nil, err
+	}
+
+	eligibleMap := make(map[uint32][]Validator)
+	waitingMap := make(map[uint32][]Validator)
+	newNodesList := make([]Validator, 0)
+
+	for _, validatorInfo := range validatorInfos {
+		chance := ihgs.nodesPerShardSetter.GetChance(validatorInfo.TempRating)
+		validator, err := NewValidator(validatorInfo.PublicKey, chance, validatorInfo.Index)
+		if err != nil {
+			return nil, err
 		}
 
-		if !willRemain {
-			actualLeaving = append(actualLeaving, shouldLeave)
+		switch validatorInfo.List {
+		case string(core.WaitingList):
+			waitingMap[validatorInfo.ShardId] = append(waitingMap[validatorInfo.ShardId], validator)
+		case string(core.EligibleList):
+			eligibleMap[validatorInfo.ShardId] = append(eligibleMap[validatorInfo.ShardId], validator)
+		case string(core.NewList):
+			newNodesList = append(newNodesList, validator)
 		}
 	}
 
-	return actualLeaving
+	sort.Slice(newNodesList, func(i, j int) bool {
```

Comment:
the indexes should be different so that no "sliceStable" or other hacks should be needed for sorting, right?

Reply:
yes. they are from genesis.